pkg/runtime/init.go

```diff
@@ -13,3 +13,263 @@
+	if err := k.LoadFromClusterfile(k.RuntimeConfig.Clusterfile); err != nil {
+		return fmt.Errorf("failed to load kubeadm config from clusterfile: %v", err)
+	}
+	// TODO handle the kubeadm config, like kubeproxy config
+
+	bs, err := k.Merge(k.getDefaultKubeadmConfig())
+	if err != nil {
+		return fmt.Errorf("failed to merge kubeadm config: %v", err)
+	}
+
+	if err := utils.WriteFile(k.getDefaultKubeadmConfig(), bs); err != nil {
+		return fmt.Errorf("failed to dump new kubeadm config: %v", err)
+	}
+	return nil
+}
+
+func (k *KubeadmRuntime) getCertSANS() []string {
+	return k.ClusterConfiguration.APIServer.CertSANs
+}
+
+//CmdToString is in host exec cmd and replace to spilt str
+func (k *KubeadmRuntime) CmdToString(host, cmd, split string) string {
+	ssh, err := k.getHostSSHClient(host)
+	if err != nil {
+		logger.Error("failed to get host ssh client, %s %v", cmd, err)
+		return ""
+	}
+	data, err := ssh.Cmd(host, cmd)
+	if err != nil {
+		logger.Error("exec remote cmd failed, %s %v", cmd, err)
+	}
+	if data != nil {
+		str := string(data)
+		str = strings.ReplaceAll(str, "\r\n", split)
+		return str
+	}
+	return ""
+}
+
+func (k *KubeadmRuntime) getRemoteHostName(hostIP string) string {
+	hostName := k.CmdToString(hostIP, "hostname", "")
+	return strings.ToLower(hostName)
+}
+
+func (k *KubeadmRuntime) GenerateCert() error {
+	err := cert.GenerateCert(
+		k.getCertPath(),
+		k.getEtcdCertPath(),
+		k.getCertSANS(),
+		k.getMaster0IP(),
+		k.getRemoteHostName(k.getMaster0IP()),
+		k.getSvcCIDR(),
+		k.getDNSDomain(),
+	)
+	if err != nil {
+		return fmt.Errorf("generate certs failed %v", err)
+	}
+	k.sendNewCertAndKey([]string{k.getMaster0IP()})
+
+	return nil
+}
+
+func (k *KubeadmRuntime) CreateKubeConfig() error {
+	hostname := k.getRemoteHostName(k.getMaster0IP())
+	certConfig := cert.Config{
+		Path:     k.getCertPath(),
+		BaseName: "ca",
+	}
+
+	controlPlaneEndpoint := fmt.Sprintf("https://%s:6443", k.getAPIServerDomain())
+	err := cert.CreateJoinControlPlaneKubeConfigFiles(k.getBasePath(),
+		certConfig, hostname, controlPlaneEndpoint, "kubernetes")
+	if err != nil {
+		return fmt.Errorf("generator kubeconfig failed %s", err)
+	}
+	return nil
+}
+
+func (k *KubeadmRuntime) CopyStaticFiles(nodes []string) error {
+	var flag bool
+	for _, file := range MasterStaticFiles {
+		staticFilePath := filepath.Join(k.getStaticFileDir(), file.Name)
+		cmdLinkStatic := fmt.Sprintf(RemoteCmdCopyStatic, file.DestinationDir, staticFilePath, filepath.Join(file.DestinationDir, file.Name))
+		var wg sync.WaitGroup
+		for _, host := range nodes {
+			wg.Add(1)
+			go func(host string) {
+				defer wg.Done()
+				ssh, err := k.getHostSSHClient(host)
+				if err != nil {
+					logger.Error("new ssh client failed %v", err)
+					flag = true
+					return
+				}
+				err = ssh.CmdAsync(host, cmdLinkStatic)
+				if err != nil {
+					logger.Error("[%s] link static file failed, error:%s", host, err.Error())
+					flag = true
+				}
+			}(host)
+			if flag {
```

Comment:
I am wondering if this line of code will never meet the condition of `flag is true`. Because when all the goroutine have been triggered and all of them have not executed to line code of `flag = true`.

Reply:
Fixed it using errChan:

```
if err != nil {
	errCh <- fmt.Errorf("[%s] link static file failed, error:%s", host, err.Error())
}
```